orttraining/orttraining/python/training/utils/hooks/_subscriber_manager.py

```diff
@@ -0,0 +1,284 @@
+    """
+
+    registered_subscribers = set()
+
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def subscribe(module: Union[torch.nn.Module, ORTModule], subscribers):
+        """
+        The API called externally to register hooks that is implicitly defined by subscribers.
+        Each time all global states will be cleaned up once called.
+        """
+        if not isinstance(module, torch.nn.Module):
+            raise ValueError("module must be a torch.nn.Module instance")
+
+        SubscriberManager._reset_all_states()
+
+        if isinstance(module, ORTModule):
+            global _EXECUTION_GLOBAL_STEP
+            # The first two steps are for preparing to export ONNX, so we use negative number for it.
+            _EXECUTION_GLOBAL_STEP -= 2
+            module = module.module
+
+        for subscriber in subscribers:
+            if not isinstance(subscriber, _ModuleHookSubscriberBase):
+                raise ValueError("subscriber must be a _ModuleHookSubscriberBase instance")
+            SubscriberManager.registered_subscribers.add(subscriber)
+
+        SubscriberManager._initialize(module)
+
+    @staticmethod
+    def _reset_all_states():
+        global _EXECUTION_GLOBAL_STEP, _OBSERVED_ACTIVATION_NAMES, _MODULE_INDEX_TO_DEPTH, _MODULE_TO_MODULE_INDEX
+        _EXECUTION_GLOBAL_STEP = 0
+        _OBSERVED_ACTIVATION_NAMES = {}
+        _MODULE_INDEX_TO_DEPTH = {}
+        _MODULE_TO_MODULE_INDEX = {}
+
+    @staticmethod
+    def _initialize(module):
+        if len(SubscriberManager.registered_subscribers) == 0:
+            warnings.warn("No subscribers are registered, skip insert hooks.")
+            return
+
+        count = [0]
+        # Register post forward hook for every module, inside the hook, we loop every tensor output of the module,
+        # and wrap it with a autograd Function called _InspectActivation (which take in a tensor and return the same
+        # tensor). In this way, we keep ORT and PyTorch run have the same boundary to check activation equality.
+        SubscriberManager._register_hooks_recursively(module, 1, count)
+
+        # Register post forward hook for outmost module, then we increase the dump step.
+        # Be noted, if backward is not triggered, the global dump step remain the original number,
+        # which means the subsequent run will override the previous dump files. This indeed happens imagine ORTModule
+        # firstly export graph (run the forward only), after gradient graph is built, another forward+backward is
+        # triggered, override the previous dump files.
+        def _post_forward_outmost_module_hook(module, inputs, outputs2):
+            def _apply_to_tensors_func(index, outputs):
+                return _IncrementStep.apply(outputs)
+
+            return SubscriberManager._apply_to_tensors(module, 0, outputs2, _apply_to_tensors_func, True)
+
+        module.register_forward_hook(_post_forward_outmost_module_hook)
+
+    @staticmethod
```

Comment:
I see quite a few staticmethods in the class. I haven't looked closely, but do they have to be static methods, or can they be moved to the module as global functions?

https://google.github.io/styleguide/pyguide.html#2174-decision